tests/test_responsetypes.py

```diff
@@ -72,17 +72,61 @@ def test_from_headers(self):
               'url': 'http://www.example.com/page/'}, Response),
+            ({'body': b'Some plain\0 text data with\0 tabs and null bytes\0'}, TextResponse),
+            ({'body': b'\x03\x02\xdf\xdd\x23', 'headers': Headers({'Content-Encoding': 'UTF-8'})},
+             Response),
+            ({'body': b'\x00\x01\xff', 'url': '://www.example.com/item/',
+              'headers': Headers({'Content-Type': ['text/plain']})}, TextResponse),
+            ({'url': 'http://www.example.com/item/file.html'}, HtmlResponse),
+            ({'body': b'<html><head><title>Hello</title></head>'}, HtmlResponse),
+            ({'body': b'<?xml version="1.0" encoding="utf-8"'}, XmlResponse),
+            ({'filename': 'file.pdf'}, TextResponse),
+            ({'url': 'http://www.example.com/item/file.pdf'}, TextResponse),
+            ({'body': b'\x01\x02', 'filename': 'file.pdf'}, Response),
+            ({'body': b'\x01\x02', 'url': 'http://www.example.com/item/file.pdf'}, Response),
+            ({'body': b'Some plain text data\1\2 with tabs and\n null bytes\0'}, Response),
+            ({'body': b'\x01\x02', 'headers': Headers({'Content-Type': ['application/pdf']})}, Response),
+            ({'headers': Headers({'Content-Type': ['application/x-json']})}, TextResponse),
+            ({'headers': Headers({'Content-Type': ['application/x-javascript']})}, TextResponse),
+            ({'headers': Headers({'Content-Type': ['application/json-amazonui-streaming']})}, TextResponse),
+        ]
+        for source, cls in mappings:
+            retcls = responsetypes.from_args(**source)
+            assert retcls is cls, f"{source} ==> {retcls} != {cls}"
 
-
+    def test_from_args_post_xtractmime(self):
+        """Each of the following test cases got affected after
+        using xtractmime for MIME sniffing"""
+        mappings = [
+            ({'body': b'\x00\x01\xff', 'url': 'http://www.example.com/item/',
+              'headers': Headers({'Content-Type': ['text/plain']})}, Response),
+            ({'filename': '/tmp/temp^'}, TextResponse),
+            ({'body': b'%PDF-1.4'}, Response),
+            ({'headers': Headers({'Content-Type': ['application/pdf']})}, TextResponse),
+            ({'headers': Headers({'Content-Type': ['application/ecmascript']})}, TextResponse),
+            ({'headers': Headers({'Content-Type': ['application/ld+json']})}, TextResponse),
+            ({'headers': Headers({'Content-Encoding': ['zip'], 'Content-Type': ['text/html']})},
+             HtmlResponse),
+            ({'headers': Headers({'Content-Encoding': ['zip'], 'Content-Type': ['text/plain']})},
+             TextResponse),
+            ({'body': b'Non HTML', 'headers': Headers({'Content-Encoding': ['zip'],
+                                                       'Content-Type': ['text/html']})}, HtmlResponse),
+            ({'body': b'Some plain text', 'headers': Headers({'Content-Type': 'application/octet-stream'})},
+             TextResponse),
```

Comment:
> > * Before, a `Content-Encoding` header would always cause a `Response`. I imagine now that’s not always the case.
> >   In fact, I suspect a non-binary body reporting some encoding before would cause the HTTPCompressionMiddleware to raise an exception, but with the new implementation the middleware should do nothing if the response is not recognized as binary.
> 
> I don’t think the tests cover this scenario. Although I might be wrong.


I think these tests are covering this scenario.

Reply:
Oops, I quotes the wrong list item :facepalm: 

> - Before, a `Content-Type` missing or for a binary type would be ignored in favor of `Content-Disposition`, URL or file name. I am guessing, for example, that with `Content-Type` `application/octet-stream` but a path-like field of a non-binary type, before you would get a non-`Response` class, and now you get `Response`.

Reply:
```
({'url': 'http://www.example.com/item/file.txt', 'headers': Headers({'Content-Type': 'application/octet-stream'})}, Response)
```
Is this the correct example for the above scenario?

Reply:
Yes, that’s what I had in mind. We might want variations with missing header and the alternatives to URL that affect the output (`Content-Disposition` and file name) for complete coverage, but that’s the base scenario.

Reply:
```
            ({'url': 'http://www.example.com/item/file.txt', 'headers': Headers({'Content-Type': 'application/octet-stream'})},
             TextResponse),
            ({'headers': Headers({'Content-Disposition': ['attachment; filename="data.xml.gz"'], 'Content-Type': 'application/octet-stream'})},
             TextResponse),
```
These tests are passing with xtractmime. I think a missing `Content-Type` or a binary content type is still ignored.

Reply:
It looks like it works for `TextResponse`. But this, for example, does not work now and I think (did not check) it would work before:

```python
({'url': 'http://www.example.com/item/file.xml', 'headers': Headers({'Content-Type': 'application/octet-stream'})}, XmlResponse),
```

Reply:
yeah you are right. It is working as a pre xtractmime test